guardian/floodgate
# Improve reindexing interactions

## What does this change?

- Disable all checkboxes on the bulk reindex page by default
- Don't remove what we _think_ may be stalled reindex ops, instead...
- Colour code in-progress and failed reindexes

Consider these changes in conjunction with https://github.com/guardian/flexible-content/pull/3527

We needed to prevent this UI from making duplicate requests for reindexes if the content source takes any time to begin working on the task. This is particularly true for Composer.

## How can we measure success?

When a Composer reindex is requested, it should return immediately and the Composer API will commence the task at some point. While we wait, this UI should just poll the content source for progress and not start making assumptions about the state of the reindex.

If a task takes some time, we will represent that as a decreasing progress bar that changes colour on the screen (from green when first started to red as it runs out of time). 

The reindex cannot be cancelled until it either starts, or the expected timeframe for starting is exceeded (currently 15 minutes).

Completed, failed, cancelled and stopped reindexes are shown with colour-coding in the reindex history panel.

## Tests
I ran the following additional checks;

- Start a reindex ✅ 
- Start the same reindex twice ✅ (Floodgate's `runningJobService` detected the existing task - no second request was sent down to the Composer API)
- Cancel a reindex ✅ 
- Run concurrent reindexes for different environments ✅  (draft-code and live-code both ran at the same time - apparently on the same Composer API instance)
- Check the activities logged by the Composer API ✅ 
- Check the activities logged by Porter ✅ (insofar as draft and live code updates were arriving and being processed) 

## Images
Here's a version of the happy path (the reindex takes a while to start but then gets going). Note that the cancel button only appears once the reindex has started.
![Floodgate_happy_path_sim](https://user-images.githubusercontent.com/690395/85547032-64353b00-b615-11ea-8a2e-b9b494518639.gif)

Here's the unhappy path (note that the cancel button only appears when the time has elapsed, but the user can still wait to see if the index gets started on its own - it isn't automatically cleared out of Floodgate's UI until it is cancelled, stopped or failed)
![Floodgate_unhappy_path_sim](https://user-images.githubusercontent.com/690395/85547080-6f886680-b615-11ea-91eb-d599935eb727.gif)

And here's the reindex history table showing us which were successful, failed, cancelled etc
![screencapture-floodgate-capi-gutools-co-uk-2020-06-24-12_19_08](https://user-images.githubusercontent.com/690395/85547129-7c0cbf00-b615-11ea-95b3-d616169d35bd.png)
